File: android/src/main/kotlin/com/haishinkit/haishin_kit/ScreenHandler.kt
package com.haishinkit.haishin_kit

import android.graphics.Rect
import android.util.Log
import com.haishinkit.screen.Screen
import com.haishinkit.screen.ScreenObject
import com.haishinkit.screen.ScreenObjectFactory
import com.haishinkit.screen.scene.ScreenObjectSnapshot
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.serialization.json.Json

class ScreenHandler(
    private var plugin: HaishinKitPlugin?,
    private var screen: Screen?
) : MethodChannel.MethodCallHandler {
    private var json = Json {
        ignoreUnknownKeys = true
    }

    private var screenObjectFactory: ScreenObjectFactory = ScreenObjectFactory()

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        when (call.method) {
            "$TAG#setSize" -> {
                val value = call.argument<String>("value")
                if (value == null) {
                    result.error(INVALID_ARGUMENTS, null, null)
                } else {
                    val size = json.decodeFromString<ScreenObjectSnapshot.Size>(value)
                    screen?.frame = Rect(0, 0, size.width, size.height)
                    result.success(null)
                }
            }
            "$TAG#addChild" -> {
                val value = call.argument<String>("value")
                if (value == null) {
                    result.error(INVALID_ARGUMENTS, null, null)
                } else {
                    val snapshot = json.decodeFromString<ScreenObjectSnapshot>(value)
                    screen?.addChild(screenObjectFactory.create(snapshot))
                    result.success(null)
                }
            }

            "$TAG#removeChild" -> {
                val value = call.argument<String>("value")
                if (value == null) {
                    result.error(INVALID_ARGUMENTS, null, null)
                } else {
                    val screenObject = screen?.findById(value)
                    screenObject?.parent?.removeChild(screenObject)
                    result.success(null)
                }
            }

            "$TAG#layout" -> {
                val value = call.argument<String>("value")
                if (value == null) {
                    result.error(INVALID_ARGUMENTS, null, null)
                } else {
                    getScreenObjectBySnapshot(value)
                    result.success(null)
                }
            }

            "$TAG#dispose" -> {
                plugin?.onDispose(hashCode())
                plugin = null
                screen = null
                result.success(null)
            }
        }
    }

    private fun getScreenObjectBySnapshot(value: String): ScreenObject {
        val snapshot = json.decodeFromString<ScreenObjectSnapshot>(value)
        val screenObject =
            screen?.findById(snapshot.id) ?: return screenObjectFactory.create(snapshot)
        screenObject.isVisible = snapshot.isVisible
        screenObject.layoutMargin.set(snapshot.layoutMargin)
        screenObject.horizontalAlignment = snapshot.horizontalAlignment
        screenObject.verticalAlignment = snapshot.verticalAlignment
        screenObject.elements = snapshot.elements
        return screenObject
    }

    private companion object {
        private const val TAG = "Screen"
        const val INVALID_ARGUMENTS = "INVALID_ARGUMENTS"
    }
}

File: ios/haishin_kit/Sources/ScreenHandler.swift
import Foundation
#if canImport(Flutter)
import Flutter
#endif
#if canImport(FlutterMacOS)
import FlutterMacOS
#endif
import HaishinKit

final class ScreenHandler: NSObject {
    enum ErrorCode: String {
        case invalidArgument = "INVALID_ARGUMENT"

        func makeFlutterError(_ message: String? = nil, details: Any? = nil) -> FlutterError {
            return FlutterError(code: self.rawValue, message: message, details: details)
        }
    }
    private var screen: Screen?
    private lazy var decoder = JSONDecoder()
    private var plugin: HaishinKitPlugin? {
        didSet {
            oldValue?.onDispose(id: Int(bitPattern: ObjectIdentifier(self)))
        }
    }

    @ScreenActor
    private lazy var screenObjectFactory = ScreenObjectFactory()

    init(plugin: HaishinKitPlugin, screen: Screen?) {
        self.plugin = plugin
        self.screen = screen
        super.init()
    }
}

extension ScreenHandler: MethodCallHandler {
    // MARK: MethodCallHandler
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let arguments = call.arguments as? [String: Any?] else {
            result(nil)
            return
        }
        switch call.method {
        case "Screen#addChild":
            guard
                let value = arguments["value"] as? String,
                let data = value.data(using: .utf8) else {
                result(ErrorCode.invalidArgument.makeFlutterError())
                return
            }
            Task { @ScreenActor in
                do {
                    let snapshot = try decoder.decode(ScreenObjectSnapshot.self, from: data)
                    let screenObject = screenObjectFactory.make(snapshot)
                    print(screenObject)
                    try screen?.addChild(screenObject)
                    result(nil)
                } catch {
                    result(ErrorCode.invalidArgument.makeFlutterError())
                }
            }
        case "Screen#removeChild":
            guard let value = arguments["value"] as? String else {
                result(ErrorCode.invalidArgument.makeFlutterError())
                return
            }
            Task { @ScreenActor in
                if let screenObject = screen?.findById(value) {
                    screenObject.parent?.removeChild(screenObject)
                }
                result(nil)
            }
        case "Screen#layout":
            guard
                let value = arguments["value"] as? String,
                let data = value.data(using: .utf8) else {
                result(ErrorCode.invalidArgument.makeFlutterError())
                return
            }
            Task { @ScreenActor in
                do {
                    try getScreenObjectBySnapshot(data)
                    result(nil)
                } catch {
                    result(ErrorCode.invalidArgument.makeFlutterError())
                }
            }
        case "Screen#dispose":
            screen = nil
            plugin = nil
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    @ScreenActor
    private func getScreenObjectBySnapshot(_ snapshot: Data) throws -> ScreenObject? {
        let snapshot = try decoder.decode(ScreenObjectSnapshot.self, from: snapshot)
        guard var screenObject = screen?.findById(snapshot.id) else {
            return screenObjectFactory.make(snapshot)
        }
        screenObject.isVisible = snapshot.isVisible
        screenObject.layoutMargin = .init(
            top: CGFloat(snapshot.layoutMargin.top),
            left: CGFloat(snapshot.layoutMargin.left),
            bottom: CGFloat(snapshot.layoutMargin.bottom),
            right: CGFloat(snapshot.layoutMargin.right)
        )
        screenObject.horizontalAlignment = .init(rawValue: snapshot.horizontalAlignment) ?? .left
        screenObject.verticalAlignment = .init(rawValue: snapshot.verticalAlignment) ?? .top
        screenObject.elements = snapshot.elements
        return screenObject
    }
}
